controllers/url_controller.js

```diff
@@ -0,0 +1,62 @@
+import Url from "../models/Url.js";
+
+const shorten_url = async (req, res, next) => {
+    try{
+        let { original_url } = req.body;
+
+        if (!original_url) {
+            return res.status(400).send({ error: 'A URL is required.' });
+        }
+
+        // Remove all whitespace from the URL and trim it
+        original_url = original_url.replace(/\s+/g, '').trim();
+
+        // Validate the URL
+        try {
+            new URL(original_url);
+        } catch (e) {
+            return res.status(400).send({ error: 'Invalid URL format.' });
+        }
+
+        const update = { original_url };
+        const options = { new: true, upsert: true, setDefaultsOnInsert: true };
+    
+        // Use findOneAndUpdate with upsert option to reduce database calls
+        const url = await Url.findOneAndUpdate({ original_url }, update, options);
```

Comment:
This 'findOneAndUpdate' function could throw an error for a duplicate key because we generate the short_url inside the Url Schema without checking if that short_url already exists in the database. We can wrap this function in a try catch and implement a mechanism for retries if the error is a duplicate key error from MongoDB.